solana/accounting-contract/programs/accounting-contract/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+use anchor_lang::prelude::*;
+use anchor_lang::system_program;
+use serde::{Deserialize, Serialize};
+
+declare_id!("31weKQJQA9ZFYaVdnjtAXUoEGPW8UUFC5TEvgPJugAub");
+
+#[program]
+pub mod accounting_contract {
+    use super::*;
+
+    pub fn set_admin(ctx: Context<SetAdmin>, new_admin: Pubkey) -> Result<()> {
+        let upgrade_authority = match ctx.accounts.program_account.upgrade_authority_address {
+            Some(address) => address,
+            None => return Err(ErrorCode::UpgradeAuthorityNotFound.into()),
+        };
+        require!(
+            ctx.accounts.caller.key() == upgrade_authority,
+            ErrorCode::Unauthorized
+        );
+        let admin = &mut ctx.accounts.admin_account.admin;
+        *admin = new_admin;
+        Ok(())
+    }
+
+    pub fn set_worker(ctx: Context<SetWorker>, new_worker: Pubkey) -> Result<()> {
+        require!(
+            ctx.accounts.admin_account.admin == ctx.accounts.caller.key(),
+            ErrorCode::Unauthorized
+        );
+        ctx.accounts.worker_account.worker = new_worker;
+        Ok(())
+    }
+
+    pub fn deposit_funds(ctx: Context<DepositFunds>, amount: u64) -> Result<()> {
+        system_program::transfer(
+            CpiContext::new(
+                ctx.accounts.system_program.to_account_info(),
+                system_program::Transfer {
+                    from: ctx.accounts.depositor.to_account_info(),
+                    to: ctx.accounts.treasury.to_account_info(),
+                },
+            ),
+            amount,
+        )?;
+        Ok(())
+    }
+
+    pub fn withdraw_funds(ctx: Context<WithdrawFunds>, amount: u64) -> Result<()> {
```

Comment:
A dumb question:
Is `ctx.accounts.beneficiary` actually part of the fn params and can be different in each call? (So that it's not firm like constructor params)

Reply:
Yes, you can set the account that is going to be used when building the transaction in the client side

Reply:
So the account cannot be part of the `fn` parameter if we're going to use it to transfer funds to it or from it, so we cannot do it via the `Pubkey`, and we need `AccountInfo`. 

In Solana, we need to be upfront all the accounts we will be using and need to pass them as part of the context for that particular instruction, so yeah `ctx.accounts.beneficiary` is part of the context constructed when calling this function which is modifiable when building the transaction.  

This is how it is constructed for reference - https://github.com/litentry/heima/blob/46bcfa001e058d87ac28b460c3692406a1439144/solana/accounting-contract/tests/src/test_initialize.rs#L107